### DISCUSSION TITLE
Little endian state encoding

### DISCUSSION THREAD

**Discussion by wcqc** [NVIDIA/cuQuantum#79] [Q&A] [answered]
It states in the doc cuQuantum uses little endian encoding. What is the best way to interface a lib that uses big endian to use cuQuantum. Qiskit uses little endian but PennyLane for example uses big endian and is integrated with cuQuantum (although looked into PennyLane's lighting.gpu code with no avail), any pointers/suggestions? There is a bit swap func in cuQuantum, can/should that be used once and for all to flip the endianess after the state is initialized? 

Thank you!

**Comment by mtjrider** [comment]
Hi @wcqc, 

This is a great question. Thanks for posting.
I assume that you are asking about how to structure qubit indices when passing them to our low-level APIs (e.g.) `custatevec.apply_matrix` ([here](https://docs.nvidia.com/cuda/cuquantum/latest/python/api/generated/cuquantum.custatevec.apply_matrix.html#cuquantum.custatevec.apply_matrix)).

Apologies in advance if this is too much detail, or not exactly what you were asking. If that is the case, please let us know.

You need not swap qubits or call another API. All you have to do is map qubit indices (like labels) to the different ordering.
I think it's easier to think about where the least/most significant index is in the qubit array:

$$ \ket{\psi} = \sum_\sigma \ket{\sigma_0 \sigma_1 \sigma_2 \cdots \sigma_{n-1}} \ \ \mathrm{w.} \ \ \sigma_i \in \\{0, 1\\} $$

Under little-endian, the most significant index corresponds with $\sigma_0$, and the least significant corresponds with $\sigma_{n-1}$.
And (to be explicit) under big-endian, $\sigma_0$ is the least significant while $\sigma_{n-1}$ is most significant.

When we say we want to apply an operator to a target qubit, $t$, we're referring to a position in the tensor product with respect to the least significant index. If $t$ assumed a big-endian qubit array, under little-endian, it would be $n-t-1$.

**Reply by wcqc** [reply]
Hi @mtjrider, 

Thanks a lot for the very detailed response, it's really helpful. The index mapping trick is clear! However I'm slightly confused about the definition of big- and little-endian in your answer though. In Qiskit for example, higher qubit indexes are more significant while in PennyLane for example lower indexes are more significant, as I understand the Qiskit scheme is little-endian while PennyLane is big-endian, which seems to be the opposite of the above? :)

Another question is it seems the endian-ness does not matter when using custatevec.initialize_state_vector (https://docs.nvidia.com/cuda/cuquantum/latest/python/api/generated/cuquantum.custatevec.initialize_state_vector.html#cuquantum-custatevec-initialize-state-vector), since the state vec is initialised directly with a "vector" not by tensoring individual qubits?

**Reply by mtjrider** [reply, marked as answer]
That is correct. State vector initialization accepts a pointer to data and does not care about endianness.

From qiskit's documentation [1]:
> Qiskit uses a slightly different ordering of the qubits, in which the qubits are represented from the most significant bit (MSB) on the left to the least significant bit (LSB) on the right (little-endian)


I'm somewhat confused by your statement regarding higher/lower qubit indices. A qubit index by itself can't be used to determine the bit-wise significance in the underlying library implementation.

A simple way to confirm bit-wise significance would be to create a 3-qubit state, and apply a single-qubit unitary on qubit 0.

If qubit 0 is most significant, the application of the operator, $u$, is equivalent to (case a):

$$ U = u \otimes I \otimes I $$

If it is least significant, it is equivalent to (case b):

$$ U = I \otimes I \otimes u $$

Big-endian corresponds with (case a) and little-endian corresponds with (case b). 


[1] Section: Basis Vector Ordering from [Summary of Quantum Operations](https://qiskit.org/documentation/tutorials/circuits/3_summary_of_quantum_operations.html)

**Reply by wcqc** [reply]
This was the page on Qiskit I was looking at, seems to be the cause of my confusion:(https://qiskit.org/documentation/apidoc/quantum_info.html)

"In Qiskit, qubits are ordered using little endian notation, with the least significant qubits having smaller indices. For example, a four-qubit system is represented as $\(|q_3q_2q_1q_0\rangle\)$. "

Just wanted to confirm again this is the same as in cuQuantum? :)

**Reply by mtjrider** [reply]
Correct! I'll post a snippet illustrating my previous comments a bit later.

**Reply by mtjrider** [reply]
Here's a sample that implements the test I described in a [previous comment](https://github.com/NVIDIA/cuQuantum/discussions/79#discussioncomment-6752213).

Note: I generated the data randomly, but the same principle applies to any gate/state.

```
import numpy as np
import functools as ft
from scipy.stats import unitary_group
from qiskit import QuantumCircuit, transpile
from qiskit_aer import AerSimulator
from qiskit.extensions import UnitaryGate


target = 0
num_qubits = 3
matrix = unitary_group.rvs(2)
gate = UnitaryGate(matrix)

initial_state = np.random.random(2**num_qubits) + 1.j * np.random.random(2**num_qubits)
initial_state = initial_state / np.linalg.norm(initial_state)

simulator = AerSimulator(method="statevector")

circuit = QuantumCircuit(num_qubits)
circuit.initialize(initial_state)
circuit = transpile(circuit, simulator)
circuit.append(gate, [target])
circuit.save_state()
result = simulator.run(circuit).result()
qiskit_state_vector = result.get_statevector().data

case_a = [matrix, np.eye(2), np.eye(2)]
case_b = [np.eye(2), np.eye(2), matrix]

operator_a = ft.reduce(lambda x, y: np.kron(x, y), case_a)
operator_b = ft.reduce(lambda x, y: np.kron(x, y), case_b)

final_state_a = operator_a @ initial_state
final_state_b = operator_b @ initial_state

if np.allclose(qiskit_state_vector, final_state_a):
    print("big-endian")
if np.allclose(qiskit_state_vector, final_state_b):
    print("little-endian")
```

**Reply by wcqc** [reply]
Thank you @mtjrider for all the detailed and super helpful responses! Will loop back with more questions if there is any.

**Reply by mtjrider** [reply]
Sure thing! Glad to help.

**Reply by mtjrider** [reply]
For posterity, here is the link in our docs clarifying bit-wise significance in our library's implementation:

https://docs.nvidia.com/cuda/cuquantum/latest/custatevec/overview.html#bit-ordering
